### ISSUE TITLE
Support PG ranges

### ISSUE THREAD

**Issue by sgrif** [diesel-rs/diesel#676]
The SQL type should be `Range<ST: CanAppearInRange>`. We should provide an impl to/from SQL for `(Bound<T>, Bound<T>) where T: FromSql<ST, Pg>`. [`Bound`](https://doc.rust-lang.org/nightly/std/collections/enum.Bound.html) is currently unstable, so we'll need to stick this behind the unstable feature. [It appears to be on track for stabilization soon though](https://github.com/rust-lang/rust/issues/27787#issuecomment-277340177), so I don't think we need to support any stable types.

For this to work with codegen we'll also need the following:

```rust
pub type Int4range = Range<Int4>;
pub type Int8range = Range<Int8>;
pub type Numrange = Range<Numeric>;
pub type Tsrange = Range<Timestamp>;
pub type Tstzrange = Range<Timestamptz>;
pub type Daterange = Range<Date>;
```

The impls for serialization/deserialization look pretty straightforward. Format appears to be a one byte set of flags (inclusive vs exclusive, whether start or end are unbounded), 4 byte length for upper value followed by data if present, 4 byte length for lower value followed by data if present.

**Comment by sgrif** [comment]
I wouldn't quite call this easy for newcomers, but it is at least fairly isolated.

**Comment by Eijebong** [comment]
I'm working on that now, I just have some questions:
- How do we treat infinity ?
- OIDs ? Apparently each range<T> has a different OID and arrays of ranges too. Should I request a metadata thing in the `CanAppearInRange` trait ?

**Comment by Eijebong** [comment]
Fixed by #1021
